### ISSUE TITLE
Applet problems are broken in Gateway Quizzes

### ISSUE THREAD

**Issue by drgrice1** [openwebwork/webwork2#1020]
This was reported in http://webwork.maa.org/moodle/mod/forum/discuss.php?d=4655.

There are several problems that prevent these sort of problems from working in gateway quizzes.

First, HEADER_TEXT is broken for gateway quizzes in general.

Second, the javascript for the addOnLoadEvent is not loaded for gateway quizzes, and that is used by the ww_applet_support.js javascript loaded in AppletObjects.pl.

Third, is some issue with named answers that I do not fully understand yet.

**Comment by drgrice1** [comment]
Actually, I think the real problem is not the named answers, but the submit action.  Problem.pm calls the submitAction javascript function anytime the problemMainForm is submitted.  The gateway quiz form does not.  The ww_applet_support.js redefines that function, and that is where the javascript values are put into the hidden inputs.

**Comment by drgrice1** [comment]
There seem to be a lot of problems here.  Even if the submitAction javascript is made to work it is not enough.  The results table is still empty.  I am not sure if this has to do with the named answers.  In any case, the named answers not having the quiz prefix will be a problem if these types of problems are used on multipage quizzes.  I am moving on from this for now.  I admit I have a bit of a lack of interest as I do not use these types of problems.

**Comment by drgrice1** [comment]
I have a fix for this now for GeogebraApplet problems.  However, there is no way to make this work for the existing problems.  Those problems will have to be changed to make them work in gateway quizzes.

So there are two ways to proceed.

The first is to make a new macro that newly written Geogebra problems should use with the way that they implement the Geogebra applet slightly modified.  The existing problems will continue to work in homework with the old macro, but will not work in gateway quizzes.

The other way is to modify the existing macro (and Applet.pm pg lib module), and require that existing problems make the neccesary changes to function with the modified macro.  This is the way that I think it should be done.  There are only a few changes that will need to be made to fix the problems.  Basically, the problems will have to start giving the applet for the problem a unique appletName (same as the appletId) and stop using names like "ggbfoo" in every problem.  No problem should ever define a global javascript ggbOnInit method.  That will be handled by code provided to the onInit option which will be called in a global ggbOnInit method defined in the ww_applet_support.js file.  Also there will be a need to take some care with the answer names.

There are only 7 problems in the OPL and 63 in Contrib that use the macro.  Almost all of the Contrib problems are @drdrew42's problems, and are also in Pending.  There are also 11 Geogebra problems in the OPL that are already not working as they attempt to use the JavaApplet object instead of the GeogebraWebApplet object.  So there is not that much that would need to be done.  I could help Andrew fix his problems, and that is most of it.

As to the other types of applet objects:

There are 87 problems that use the FlashApplet object in the OPL, 3 in Contrib, and those same 3 in Pending.  However, flash support is being removed from all browsers in general, and shouldn't be used.

I already mentioned the 11 JavaApplet problems that are really Geogebra problems, and they are the only ones in the OPL.  Java support in browsers is also fading.

Finally, the CanvasObject.  There are no problems that use this.  Of course canvas support is growing, so this should perhaps be kept.  If anyone has any problems that use this, I could try to make sure this still functions if the existing macro is renovated.

I am posting this, as I am looking for feedback as to how we want to proceed.

**Comment by drdrew42** [comment]
My main concern here is for those WeBWorK servers with custom content that has not been pushed back into the OPL. For that reason, I think the first solution is preferable -- though with such a solution, I would prefer to update all of my GeoGebra problems to use the new macro. I also have an interest in making sure that the evolution of a geogebra macro would support my existing problems, some of which get pretty hack-y, and allow me to further  wreak havoc.

The most important thing, IMO, with such a solution would be to update the documentation so that content authors migrate away from the old macro. For example, we still have authors using the old code snippet with "geogebra.jar" even though we haven't supported the use of that implementation for several years now.

**Comment by drgrice1** [comment]
In my opinion it is the responsibility of those that have custom content not publicly available to update their problems to work with any changes that are made.  I do not agree with imposing a backwards compatibility restriction for that reason.  Such restrictions are a major hindrance to forward progress for WeBWorK.  This causes an indefinite blockade to deprecation as there is no way to no what is out there.  That is why there is deprecated code in WeBWorK that has been marked as such for more than a decade.  

Furthermore, in this case the changes that will be needed will be minor.  In fact there is only one thing that would need to be changed to make it work for homework again (see below).

The 11 problems in the OPL I mentioned that use the JavaApplet object use the "geogebra.jar" approach that you mentioned.  No author should be using that, and all support for that approach should be dropped.  If it is used, students will have trouble with it.  I said in my last comment that Java support is fading, but Java support is further gone than Flash support.  Firefox and Chrome don't support it at all anymore.  Neither does Safari.  It is hard to find a browser that does.

I believe the only applets we should support going forward are geogebra and canvas applets.  They are based on JavaScript and HTML5 which will have support from browsers for the foreseeable future.

In any case, I am trying to find a third option that will fix the existing macro without creating a new one, but still allow for the problems like yours to continue to work in homework.  Although they still would need modification to work in Gateway quizzes.  The one thing that is difficult to overcome is the ggbOnInit issue.  Geogebra's web deployment approach relies on a global ggbOnInit function.  There can be only one!  That one should be one that is defined in the ww_applet_support.js file.  That function should call the onInit method for each applet (which should not be named ggbOnInit).  The problem is that if it is defined there, than a ggbOnInit function defined in a problem is overridden (or vice versa depending on the order things appear in the DOM).  This is the one thing that may need to be changed to make existing problems work in homework if I can't find a work around.  Change the name of a ggbOnInit function to something else, and make sure that it is set for the onInit option.  (Note that the onInit option at this point could be set to 1 or anything that perl evaluates to true and the ggbOnInit method must be named ggbOnInit regardless of what the onInit option is.)

**Comment by drgrice1** [comment]
I have the third option working now.  Problems that are written in the old way will still work in homework, and problems written in a slightly modified new way will work in both homework and gateway quizzes.  One downside is that problems written in the new correct way, will cease to work with versions of webwork prior to the next release.

To fix problems to work in gateway quizzes, authors will need to do three things.
1.  Change the name of any ggbOnInit method defined to something else, or even better just pass in javascript code to onInit to be execute by the main ggbOnInit method.  The latter approach avoids the possibility of a function name clash with another problem.  The updated javascript code is designed to work with either though.
2.  Make sure the appletName (same as the appletId) is something that is generally unique.  At least try to make it different than that of other problems.  If two problems have the same appletName, they won't work on the same page.
3.  Use a different approach to naming (labeling) answers.  I will probably try to add some convenience methods to the macro to make this easier for authors to work with.

I will put in a pull request with the fixes for applets soon.  I just need to add some documentation for the new way to code problems, and add some named answer convenience methods.

Generally the naming issue is something that needs to be done with named answers for all problems to make them work correctly in gateway quizzes.  Currently many have taken the approach of creating a name for the answer rule if it is needed in the problem (for something like JavaScript or to change the order answer rules are processed).  That is incorrect.  The correct approach is to request an answer name from PG using the NEW_ANS_NAME method and use that.  There are some work arounds that are needed to get the QUIZ_PREFIX to work right in some cases, and this is something that we will need to redesign to make this more convenient for authors.  If this is done it will fix the PG issue https://github.com/openwebwork/pg/issues/184 in the best possible way.

The current workaround that is needed is that an answer name provided by NEW_ANS_NAME needs to have the QUIZ_PREFIX stripped off before that name is passed to NAMED_ANS (or LABELED_ANS) .  The problem is that NEW_ANS_NAME returns an answer name that has the QUIZ_PREFIX already prefixed, and then NAMED_ANS again adds the QUIZ_PREFIX.  Resulting in an error like the answer rule Q0001_AnSwEr0001 has no answer evaluator and the answer evaluatore Q0001_Q0001_AnSwEr0001 has no answer rule.

Currently problems that use the approach of creating a name for the answer rule don't function in gateway quizzes, as the answer rule does not have the QUIZ_PREFIX, and the NAMED_ANS method adds it so that the answer evaluator does.  Resulting in a similar error to above.  All answers in gateway quizzes must contain the QUIZ_PREFIX to function correctly.

**Comment by drgrice1** [comment]
I have submitted a pull request that makes it possible for applet problems to work in gateway quizzes.  See https://github.com/openwebwork/pg/pull/489 and https://github.com/openwebwork/webwork2/pull/1121.

**Comment by drdrew42** [comment]
I did a first pass review, and things look good. I like the redesign -- particularly the incorporation of ADD_JS_FILE and ADD_CSS_FILE into PG. 

I will have to do some targeted testing with my GGB problems that use a subclass of MultiAnswer.
(e.g. webwork-open-problem-library/Contrib/CUNY/CityTech/CollegeAlgebra_Trig/setGeogebra/circle-center-multians-inc/circle-center-multians-inc-test.pg uses a hard-coded reference to 'answerBox' ... though it is definitely not a problem that should be used in a gateway quiz setting in the first place.) Not sure when I'll be able to get to that, perhaps later this weekend.

**Comment by drgrice1** [comment]
Actually, I didn't add the ADD_CSS_FILE method.  @taniwallach did that in https://github.com/openwebwork/webwork2/pull/1018 and https://github.com/openwebwork/pg/pull/428.  I tweaked the code a bit.  There is no need to check if the hash key extra_css_files exists.  Perl nicely deals with creating it as an array when you push to it.  In any case, both methods are useful.

I have a version of one of your problems that I used in testing that is modified to work in both homework and gateway quizzes.  I removed the MultiAnswer subclass, but other than that made only the minimal changes needed to make it work.  I will email it to you.
